File: backend/internal/trackers/schema.go
package trackers

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
)

type Schema struct {
	Fields []FieldDef `json:"fields"`
}

type FieldDef struct {
	Key           string            `json:"key"`
	Label         map[string]string `json:"label"`
	Type          string            `json:"type"`
	Required      bool              `json:"required,omitempty"`
	IsPrimaryTime bool              `json:"isPrimaryTime,omitempty"`

	Unit    string        `json:"unit,omitempty"`
	Min     *float64      `json:"min,omitempty"`
	Max     *float64      `json:"max,omitempty"`
	Options []FieldOption `json:"options,omitempty"`
}

type FieldOption struct {
	Value string            `json:"value"`
	Label map[string]string `json:"label"`
}

var (
	keyRe = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)

	allowedTypes = map[string]struct{}{
		"datetime":    {},
		"date":        {},
		"time":        {},
		"duration":    {},
		"number":      {},
		"text":        {},
		"longtext":    {},
		"select":      {},
		"multiselect": {},
		"boolean":     {},
		"color":       {},
	}

	timeTypes = map[string]struct{}{
		"datetime": {},
		"date":     {},
		"time":     {},
	}
)

// ParseSchema parses raw JSON, validates it, and returns the canonical form.
func ParseSchema(raw []byte) (*Schema, error) {
	if len(raw) == 0 {
		return nil, errors.New("schema_json is empty")
	}
	var s Schema
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&s); err != nil {
		return nil, fmt.Errorf("decode schema: %w", err)
	}
	if err := ValidateSchema(&s); err != nil {
		return nil, err
	}
	return &s, nil
}

// MarshalSchema produces canonical JSON for storage.
func MarshalSchema(s *Schema) ([]byte, error) { return json.Marshal(s) }

func ValidateSchema(s *Schema) error {
	if s == nil {
		return errors.New("schema is nil")
	}
	if len(s.Fields) == 0 {
		return errors.New("schema must have at least one field")
	}
	seenKeys := map[string]struct{}{}
	primaryCount := 0
	for i := range s.Fields {
		f := &s.Fields[i]
		if !keyRe.MatchString(f.Key) {
			return fmt.Errorf("field[%d]: key %q must match [a-z][a-z0-9_]*", i, f.Key)
		}
		if _, dup := seenKeys[f.Key]; dup {
			return fmt.Errorf("field[%d]: duplicate key %q", i, f.Key)
		}
		seenKeys[f.Key] = struct{}{}
		if f.Label == nil || f.Label["en"] == "" {
			return fmt.Errorf("field[%d] (%s): label.en is required", i, f.Key)
		}
		if _, ok := allowedTypes[f.Type]; !ok {
			return fmt.Errorf("field[%d] (%s): unknown type %q", i, f.Key, f.Type)
		}
		if f.IsPrimaryTime {
			if _, ok := timeTypes[f.Type]; !ok {
				return fmt.Errorf("field[%d] (%s): isPrimaryTime requires type datetime|date|time", i, f.Key)
			}
			primaryCount++
		}
		switch f.Type {
		case "number":
			if f.Min != nil && f.Max != nil && *f.Min > *f.Max {
				return fmt.Errorf("field[%d] (%s): min > max", i, f.Key)
			}
		case "select", "multiselect":
			if len(f.Options) == 0 {
				return fmt.Errorf("field[%d] (%s): %s requires non-empty options", i, f.Key, f.Type)
			}
			seenVals := map[string]struct{}{}
			for j, opt := range f.Options {
				if opt.Value == "" {
					return fmt.Errorf("field[%d] (%s): option[%d] value is required", i, f.Key, j)
				}
				if _, dup := seenVals[opt.Value]; dup {
					return fmt.Errorf("field[%d] (%s): duplicate option value %q", i, f.Key, opt.Value)
				}
				seenVals[opt.Value] = struct{}{}
				if opt.Label == nil || opt.Label["en"] == "" {
					return fmt.Errorf("field[%d] (%s): option[%d] label.en is required", i, f.Key, j)
				}
			}
		}
		if f.Type != "number" && (f.Min != nil || f.Max != nil || f.Unit != "") {
			return fmt.Errorf("field[%d] (%s): min/max/unit only valid for number", i, f.Key)
		}
		if f.Type != "select" && f.Type != "multiselect" && len(f.Options) > 0 {
			return fmt.Errorf("field[%d] (%s): options only valid for select/multiselect", i, f.Key)
		}
	}
	if primaryCount > 1 {
		return errors.New("at most one field may have isPrimaryTime: true")
	}
	return nil
}

// PrimaryTimeKey returns the key of the field marked isPrimaryTime, or "" if none.
func (s *Schema) PrimaryTimeKey() string {
	if s == nil {
		return ""
	}
	for _, f := range s.Fields {
		if f.IsPrimaryTime {
			return f.Key
		}
	}
	return ""
}

File: frontend/src/lib/schema.ts
export type FieldType =
  | 'datetime'
  | 'date'
  | 'time'
  | 'duration'
  | 'number'
  | 'text'
  | 'longtext'
  | 'select'
  | 'multiselect'
  | 'boolean'
  | 'color'

export const FIELD_TYPES: FieldType[] = [
  'datetime',
  'date',
  'time',
  'duration',
  'number',
  'text',
  'longtext',
  'select',
  'multiselect',
  'boolean',
  'color',
]

export const TIME_FIELD_TYPES: FieldType[] = ['datetime', 'date', 'time']

export interface FieldOption {
  value: string
  label: { en: string; uk?: string }
}

export interface FieldDef {
  key: string
  label: { en: string; uk?: string }
  type: FieldType
  required?: boolean
  isPrimaryTime?: boolean
  unit?: string
  min?: number
  max?: number
  options?: FieldOption[]
}

export interface Schema {
  fields: FieldDef[]
}

export const KEY_REGEX = /^[a-z][a-z0-9_]*$/

export function defaultFieldByType(type: FieldType): FieldDef {
  const base: FieldDef = {
    key: '',
    label: { en: '' },
    type,
  }
  if (type === 'select' || type === 'multiselect') {
    base.options = [{ value: '', label: { en: '' } }]
  }
  return base
}

export function validateSchemaClient(schema: Schema): string[] {
  const errors: string[] = []
  if (!schema.fields.length) {
    errors.push('At least one field is required')
    return errors
  }
  const seenKeys = new Set<string>()
  let primary = 0
  schema.fields.forEach((f, idx) => {
    const where = `Field #${idx + 1}${f.key ? ` (${f.key})` : ''}`
    if (!KEY_REGEX.test(f.key)) {
      errors.push(`${where}: key must match [a-z][a-z0-9_]*`)
    }
    if (seenKeys.has(f.key)) {
      errors.push(`${where}: duplicate key "${f.key}"`)
    } else if (f.key) {
      seenKeys.add(f.key)
    }
    if (!f.label?.en) {
      errors.push(`${where}: label.en is required`)
    }
    if (!FIELD_TYPES.includes(f.type)) {
      errors.push(`${where}: unknown type "${f.type}"`)
    }
    if (f.isPrimaryTime) {
      if (!TIME_FIELD_TYPES.includes(f.type)) {
        errors.push(`${where}: isPrimaryTime requires datetime|date|time`)
      }
      primary++
    }
    if (f.type === 'select' || f.type === 'multiselect') {
      if (!f.options?.length) {
        errors.push(`${where}: ${f.type} requires at least one option`)
      } else {
        const seenVals = new Set<string>()
        f.options.forEach((opt, i) => {
          if (!opt.value) errors.push(`${where}: option #${i + 1} value is required`)
          if (seenVals.has(opt.value)) errors.push(`${where}: duplicate option "${opt.value}"`)
          else if (opt.value) seenVals.add(opt.value)
          if (!opt.label?.en) errors.push(`${where}: option #${i + 1} label.en is required`)
        })
      }
    }
  })
  if (primary > 1) errors.push('At most one field may have isPrimaryTime')
  return errors
}

export function emptySchema(): Schema {
  return { fields: [] }
}
